vault/expiration.go

```diff
@@ -245,132 +290,94 @@ func (m *ExpirationManager) Restore() error {
-					result <- le
-
-				// quit early
-				case <-quit:
-					return
-				}
-			}
-		}()
-	}
-
-	// Distribute the collected keys to the workers in a go routine
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		for i, leaseID := range existing {
-			if i%500 == 0 {
-				m.logger.Trace("expiration: leases loading", "progress", i)
-			}
-
-			select {
-			case <-quit:
-				return
-
-			default:
-				broker <- leaseID
-			}
-		}
-
-		// Close the broker, causing worker routines to exit
-		close(broker)
-	}()
-
 	// Restore each key by pulling from the result chan
-	for i := 0; i < len(existing); i++ {
-		select {
-		case err := <-errs:
-			// Close all go routines
-			close(quit)
+	restoredCount := 0
+	for i, leaseID := range existing {
+		if i > 0 && i%500 == 0 {
+			m.logger.Trace("expiration: leases loading", "progress", i)
+		}
 
+		le, err := m.loadAndRestoreLease(leaseID)
+		if err != nil {
 			return err
+		}
+		if le != nil {
+			restoredCount++
+		}
 
-		case le := <-result:
+		// This exists for test only to add latency to load calls
+		if loadDelay > 0 {
+			time.Sleep(loadDelay)
+		}
+	}
 
-			// If there is no entry, nothing to restore
-			if le == nil {
-				continue
-			}
+	// Clear our the restored entries and turn off restore mode
+	m.restoreMutex.Lock()
+	m.restoreLoaded = make(map[string]struct{})
+	atomic.StoreInt64(&m.restoreMode, 0)
+	m.restoreMutex.Unlock()
 
-			// If there is no expiry time, don't do anything
-			if le.ExpireTime.IsZero() {
-				continue
-			}
+	m.logger.Info("expiration: leases restored", "restored_lease_count", restoredCount)
+	return nil
+}
 
-			// Determine the remaining time to expiration
-			expires := le.ExpireTime.Sub(time.Now())
-			if expires <= 0 {
-				expires = minRevokeDelay
-			}
+func (m *ExpirationManager) loadAndRestoreLease(leaseID string) (*leaseEntry, error) {
+	m.restoreMutex.Lock()
+	defer m.restoreMutex.Unlock()
 
-			// Setup revocation timer
-			m.pending[le.LeaseID] = time.AfterFunc(expires, func() {
-				m.expireID(le.LeaseID)
-			})
-		}
+	// If we loaded a lease elsewhere in the expiration manager while in
+	// restore, skip the entry since it may have been updated since we
+	// loaded it
+	if _, ok := m.restoreLoaded[leaseID]; ok {
+		return nil, nil
 	}
 
-	// Let all go routines finish
-	wg.Wait()
+	le, err := m.loadEntryOnly(leaseID)
+	if err != nil {
+		return nil, err
+	}
 
-	if len(m.pending) > 0 {
-		if m.logger.IsInfo() {
-			m.logger.Info("expire: leases restored", "restored_lease_count", len(m.pending))
-		}
+	m.restoreLease(le)
+	return le, nil
+}
+
+// restoreLease takes a lease entry that has not been added to the expiration
+// manager and adds it back in
+func (m *ExpirationManager) restoreLease(le *leaseEntry) {
+	// If there is no entry, nothing to restore
+	if le == nil {
+		return
 	}
 
-	return nil
+	// Update the cache of restored leases, either synchronously or through
+	// the lazy loaded restore process
+	m.restoreLoaded[le.LeaseID] = struct{}{}
```

Comment:
I thought the map was not going to hold an exhaustive list but rather only the list of leases that were restored sync. A map can handle a million entries, sure, so it's not necessarily a problem this way.

Reply:
This ended up being the simpler implementation.  What we didn't want to do is to populate the map from collecting keys and popping things off that map.  This is just keeping a record of things we have seen not what we still need to process in the restore loop.